lib/services/downloads_service_backend.dart

```diff
@@ -0,0 +1,1515 @@
+  final _finampUserHelper = GetIt.instance<FinampUserHelper>();
+
+  IsarTaskQueue(this._downloadsService);
+
+  /// Set of tasks that are believed to be actively running
+  final _activeDownloads = <int>{}; // by TaskId
+
+  Completer<void>? _callbacksComplete;
+
+  final _isar = GetIt.instance<Isar>();
+
+  /// Initialize the queue and start stored downloads.
+  /// Should only be called after background_downloader and downloadsService are
+  /// fully set up.
+  Future<void> initializeQueue() async {
+    _activeDownloads.addAll(
+        (await FileDownloader().allTasks(includeTasksWaitingToRetry: true))
+            .map((e) => int.parse(e.taskId)));
+    List<DownloadItem> completed = [];
+    List<DownloadItem> needsEnqueue = [];
+    for (var item in _isar.downloadItems
+        .where()
+        .stateEqualTo(DownloadItemState.enqueued)
+        .or()
+        .stateEqualTo(DownloadItemState.downloading)
+        .filter()
+        .typeEqualTo(DownloadItemType.song)
+        .or()
+        .typeEqualTo(DownloadItemType.image)
+        .findAllSync()) {
+      if (item.file?.existsSync() ?? false) {
+        _activeDownloads.remove(item.isarId);
+        completed.add(item);
+      } else if (item.state == DownloadItemState.downloading) {
+        if (!_activeDownloads.contains(item.isarId)) {
+          needsEnqueue.add(item);
+        }
+      }
+    }
+    _isar.writeTxnSync(() {
+      // Images marked as completed this way will not recieve updated extensions like ones
+      // processed in status updates, but that's not really important
+      for (var item in completed) {
+        _downloadsService.updateItemState(item, DownloadItemState.complete);
+        _enqueueLog
+            .info("Marking download ${item.name} as complete on startup.");
+      }
+      for (var item in needsEnqueue) {
+        _downloadsService.updateItemState(item, DownloadItemState.enqueued);
+        _enqueueLog.info("Re-enqueueing download ${item.name} on startup.");
+      }
+    });
+  }
+
+  /// Execute all pending downloads.
+  Future<void> executeDownloads() async {
+    if (_callbacksComplete != null) {
+      return _callbacksComplete!.future;
+    }
+    try {
+      _callbacksComplete = Completer();
+      unawaited(_advanceQueue());
+      await _callbacksComplete!.future;
+      _enqueueLog.info("All downloads enqueued.");
+    } finally {
+      _callbacksComplete = null;
+    }
+  }
+
+  /// Advance the queue if possible and ready, no-op if not.
+  /// Will loop until all downloads have been enqueued.  Will enqueue
+  /// finampSettings.maxConcurrentDownloads at once.
+  Future<void> _advanceQueue() async {
+    try {
+      while (true) {
+        var nextTasks = _isar.downloadItems
+            .where()
+            .stateEqualTo(DownloadItemState.enqueued)
+            .filter()
+            .allOf(_activeDownloads,
+                (q, element) => q.not().isarIdEqualTo(element))
+            .limit(20)
```

Comment:
What is this limit for? Shouldn't it be tied to maxConcurrentDownloads or a similar setting?

Reply:
This limit controls how batched the database calls are.  Decreasing it means more database reads, increasing it means a slightly higher chance of downloading a deleted item or something.  It really doesn't matter that much, 20 just seemed like a decent balance.